net/http/src/THttpLongPollEngine.cxx

```diff
@@ -79,37 +154,40 @@ void THttpLongPollEngine::SendCharStar(const char *buf)
 /// function called in the user code before processing correspondent websocket data
 /// returns kTRUE when user should ignore such http request - it is for internal use
 
-Bool_t THttpLongPollEngine::PreviewData(THttpCallArg &arg)
+Bool_t THttpLongPollEngine::PreviewData(std::shared_ptr<THttpCallArg> &arg)
 {
-   if (!strstr(arg.GetQuery(), "&dummy")) {
+   if (!strstr(arg->GetQuery(), "&dummy")) {
       // this is normal request, deliver and process it as any other
       // put dummy content, it can be overwritten in the future
-      arg.SetContentType("text/plain");
-      arg.SetContent(gLongPollNope);
+      arg->SetTextContent(std::string(gLongPollNope));
       return kFALSE;
    }
 
-   if (&arg == fPoll) {
-      Error("PreviewData", "NEVER SHOULD HAPPEN");
-      gSystem->Exit(12);
-   }
+   if (arg == fPoll)
+      Fatal("PreviewData", "Submit same THttpCallArg object once again");
 
    if (fPoll) {
-      Info("PreviewData", "Get dummy request when previous not completed");
+      Error("PreviewData", "Get next dummy request when previous not completed");
       // if there are pending request, reply it immediately
-      fPoll->SetContentType("text/plain");
-      fPoll->SetContent(gLongPollNope); // normally should never happen
-      fPoll->NotifyCondition();
-      fPoll = nullptr;
+      if (fRaw) fPoll->SetBinaryContent(std::string("txt:") + gLongPollNope);
+           else fPoll->SetTextContent(std::string(gLongPollNope)); // normally should never happen
+      fPoll->NotifyCondition();         // inform http server that request is processed
+      fPoll.reset();
    }
 
    if (fQueue.size() > 0) {
-      arg.SetContentType("text/plain");
-      arg.SetContent(fQueue.front().c_str());
+      QueueItem &item = fQueue.front();
+      if (item.fBinary) {
+         arg->SetBinaryContent(std::move(item.fData));
+         if (!fRaw && !item.fHdr.empty())
+            arg->SetExtraHeader("LongpollHeader", item.fHdr.c_str());
+      } else {
+         arg->SetTextContent(std::move(item.fData));
+      }
       fQueue.erase(fQueue.begin());
```

Comment:
Should `fQueue` be a `std::queue`? `pop_front()`?

Reply:
Done